### PR TITLE
add SNXRecoveryCoordinator

### PR DESCRIPTION
**Pull request by TomAFrench**
This PR adds the SNXRecoveryCoordinator. This coordinator is designed to return the SNX, sBTC and sUSD drained from the Vault for protection while the double-entrypoint issue was being fixed by the Synthetix team.

After this coordinator is run the ProtocolFeeWithdrawer will treat these tokens as regular tokens so they will be able to be withdrawn by the Treasury subDAO as usual.

### CODE REVIEW

**Comment by TomAFrench** [comment]
tl;dr: These are the amounts which need to be sent to the Vault in order to make it whole.
|  Network | Token | amount|
| ------------ | -------  | ------- |
| mainnet | SNX |  937727.16385483176745 |
| mainnet | sBTC | 273.030307592426881329  |
| optimism | SNX | 5,169.589929780230696166  |
| optimism | sUSD |  22143.443690237097992 |

**Calculation**

In order to return the proper amount of SNX and sBTC to the Vault we need to find all instances of flashloans which took advantage of the double-entrypoint mechanism and total up the amount sent to the PFC. We can do this by just looking at transfers from the Vault to the PFC. 

# Mainnet

## SNX

https://etherscan.io/token/0xc011a73ee8576fb46f5e1c5751ca3b9fe0af2a6f?a=0xce88686553686da562ce7cea497ce749da109f9f

image

There are two transactions of interest here:
1. https://etherscan.io/tx/0xff2f3cc8322e8ed72966dba664938286981fb3cdf9ac3b29c06b3a0ff4769f4f
2. https://etherscan.io/tx/0x13a07548dadc6c7465dc805fdf42bb45afe7ba1d8c2d21bff86dd56ec4409d67

These resulted in 937,711.95835226019046332 + 15.205502571576986197 = 937727.16385483176745 SNX being sent to the PFC. All other transfers are standard collections of protocol fees and can be ignored.

We then need to send 937727.16385483176745 SNX to the vault from the PFC.

## sBTC

https://etherscan.io/token/0xfe18be6b3bd88a2d2a7f928d00292e7a9963cfc6?a=0xce88686553686da562ce7cea497ce749da109f9f

image

The only transaction of interest here is:
1. https://etherscan.io/tx/0xff2f3cc8322e8ed72966dba664938286981fb3cdf9ac3b29c06b3a0ff4769f4f

We then need to send 273.030307592426881329 sBTC to the vault from the PFC.

# Optimism

## SNX

https://optimistic.etherscan.io/token/0x8700daec35af8ff88c16bdf0418774cb3d7599b4?a=0xce88686553686da562ce7cea497ce749da109f9f

image

The only transaction of interest here is:
1. https://optimistic.etherscan.io/tx/0x79bd59b72918a0b1143b504aacf05efd98a5572597ff6e9c5bb17d71ef6a7282

We then need to send 5,169.589929780230696166 SNX to the vault from the PFC.

## sUSD

https://optimistic.etherscan.io/token/0x8c6f28f2f1a3c87f0f938b96d27520d9751ec8d9?a=0xce88686553686da562ce7cea497ce749da109f9f

image

There are two transactions of interest here:
1. https://optimistic.etherscan.io/tx/0xd3ef3a65dc3c3423a888f121b88e359a4932ee00d57fd1f55404e5266316fd81
2. https://optimistic.etherscan.io/tx/0x7e29b4277188c6c80debe499752d1674545a914037bee674e9ec214de7f31d6a

These resulted in 7,123.674988883992841125 + 15,019.768701353105150733 = 22143.443690237097992 sUSD being sent to the PFC.

We then need to send 22143.443690237097992 sUSD to the vault from the PFC.

**Comment by nventuro** [comment]
I've checked the build info, token addresses, transactions and amounts (thanks for the detailed writeup!).

Everything seems to be alright, except there's a small rounding issue in the additions (not sure where it comes from). E.g. the sUSD amounts are 7,123.674988883992841125 and 15,019.768701353105150733, but you listed the result as 22143.443690237097992. The correct result is 22143.443690237097991858 (see how the smallest decimals are 5 and 3, so the result should have a smallest decimal of 8, not 2).

Similarly for mainnet SNX, the correct result is 93,772.7163854831767449517.

**Comment by TomAFrench** [comment]
Ah, looks like I had some rounding creep in. Checked the maths now and I get the same as you.

image
